src/unity.c

```diff
@@ -235,97 +235,121 @@ void UnityPrintMask(const UNITY_UINT mask, const UNITY_UINT number)
+/* If rounds up && remainder 0.5 && result odd && below cutoff for double precision issues */
+  #define ROUND_TIES_TO_EVEN(orig, num_int, num)                                          \
+  if (num_int > (num) && (num) - (num_int-1) <= 0.5 && (num_int & 1) == 1 && orig < 1e22) \
+    num_int -= 1 /* => a tie to round down to even */
+#else
+  #define ROUND_TIES_TO_EVEN(orig, num_int, num) /* Remove macro */
+#endif
+
+/* Printing floating point numbers is hard. Some goals of this implementation: works for embedded
+ * systems, floats or doubles, and has a reasonable format. The key paper in this area,
+ * 'How to Print Floating-Point Numbers Accurately' by Steele & White, shows an approximation by
+ * scaling called Dragon 2. This code uses a similar idea. The other core algorithm uses casts and
+ * floating subtraction to give exact remainders after the decimal, to be scaled into an integer.
+ * Extra trailing 0's are excluded. The output defaults to rounding to nearest, ties to even. You
+ * can enable rounding ties away from zero. Note: UNITY_DOUBLE param can typedef to float or double
+
+ * The old version required compiling in snprintf. For reference, with a similar format as now:
+ *  char buf[19];
+ *  if (number > 4294967296.0 || -number > 4294967296.0) snprintf(buf, sizeof buf, "%.8e", number);
+ *  else if (number < 0.1 && number > -0.1)              snprintf(buf, sizeof buf, "%.5e", number);
+ *  else                                                 snprintf(buf, sizeof buf, "%.6f", number);
+ *  UnityPrint(buf);
+ */
 void UnityPrintFloat(const UNITY_DOUBLE input_number)
 {
     UNITY_DOUBLE number = input_number;
 
     /* print minus sign (including for negative zero) */
-    if (number < 0.0f || (number == 0.0f && 1.0f / number < 0.0f))
+    if (number < 0 || (number <= 0 && 1.0f / number < 0))
     {
         UNITY_OUTPUT_CHAR('-');
         number = -number;
     }
 
-    /* handle zero, NaN, and +/- infinity */
-    if (number == 0.0f) UnityPrint("0");
-    else if (isnan(number)) UnityPrint("nan");
-    else if (isinf(number)) UnityPrint("inf");
-    else
+    if (isnan(number)) UnityPrint(UnityStrNaN);
+    else if (isinf(number)) UnityPrintLen(UnityStrInf, 3);
+    else if (number < 0.1 && number > 0)
     {
+        /* Prints only 6 significant digits. */
         int exponent = 0;
-        int decimals, digits;
         UNITY_INT32 n;
-        char buf[16];
 
-        /* scale up or down by powers of 10 */
+        /* scale up by powers of 10 */
         while (number < 100000.0f / 1e6f)  { number *= 1e6f; exponent -= 6; }
         while (number < 100000.0f)         { number *= 10.0f; exponent--; }
-        while (number > 1000000.0f * 1e6f) { number /= 1e6f; exponent += 6; }
-        while (number > 1000000.0f)        { number /= 10.0f; exponent++; }
 
         /* round to nearest integer */
-        n = ((UNITY_INT32)(number + number) + 1) / 2;
+        n = (UNITY_INT32)(number + 0.5f);
```

Comment:
Doing the rounding in integer math was intentional, because an int32 has more precision than a float.  If you do it in float, you can completely lose the 0.5f due to truncation and get incorrect rounding in some cases.

Reply:
That's only true outside the range of numbers we are dealing with here. `100,000` to `1,000,000` can always add `0.5f` within the precision. But you knew that ( I can tell you know a thing or two about floats).
I thought this code was cleaner, showing more intent. It's a judgement call.

Reply:
Hmm, yes, this mattered for the version I had locally that was printing 7 digits but doesn't matter for 6.  Good point.